pkg/overlay/client.go

```diff
@@ -41,6 +40,9 @@ func NewOverlayClient(address string) (*Overlay, error) {
 	}, nil
 }
 
+// a compiler trick to make sure *Overlay implements Client
+var _ Client = (*Overlay)(nil)
```

Comment:
Wait, what? Can you explain why this is needed ?

The compiler will complain if Overlay does not implement Client when a user of this service that imports `Overlay.Client` attempts to use `NewOverlayClient`. We want it to return the concrete type and not an interface because that's what we will be using. It implements the interface but we will be using the concrete type.

Reply:
Yes, changes my first concern. 

My second is point is unchanged though, this should return the concrete type not the interface.

Reply:
oh, maybe I missed something. where are we returning an interface and not a concrete type?

Reply:
We aren't, I was responding to @kaloyan-raev 's comment

Reply:
I am really confused that a constructor with name pattern `New<InterfaceName>()` would return anything else but the `<InterfaceName>` interface.

What's the use case that the caller of `New<InterfaceName>()` would expect anything different?

Reply:
The interface is Client, the concrete implementation is called Overlay. this is a pretty common naming pattern in the std lib.

A constructor always returns a a concrete implementation of an interface (if there is a counter example I'm not aware of one).

https://golang.org/pkg/io/#SectionReader Implements `Reader` interface and the New method returns the concrete impl.

Reply:
OK. But you also have this: https://golang.org/pkg/io/#LimitReader

It returns `io.Reader` instead of `*io.LimitedReader`. We have similar examples in our ranger packages too.

I just wonder how one decides on the approach.

Reply:
I'm not sure why they made the decision, so yes that would be a good thing to figure out but, I think they made it very clear what was happening here, both not calling it `NewLimitReader` and calling it out in their comment.
```
LimitReader returns a Reader that reads from r but stops with EOF after n bytes. The underlying implementation is a *LimitedReader. 
```